Env config

THardy98 opened the pull request:
## What was changed
Add environment configuration feature

1. Closes #490

2. How was this tested:
Integration test suite `ClientConfigTests.cs`

3. Any docs updates needed?
Yes

THardy98 committed:
Env config

THardy98 commented:
Relies on https://github.com/temporalio/sdk-core/pull/986

THardy98 committed:
remove merge conflicts

cretz commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
```suggestion
namespace Temporalio.Client.EnvConfig
```

I think `EnvConfig` is better because 1) that's what we're calling it everywhere else, and 2) there's a lot of configuration not in here, so it's not really general purpose "configuration"

cretz commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
```suggestion
    public sealed record ClientConfig(IReadOnlyDictionary<string, ClientConfigProfile> profiles)
    {
```
I think this is better as a record instead of a simple class like our "options" classes are. And can remove cloneable since records have their own way to kinda clone. Same for the other options.

cretz commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
For many languages, this is generic "client config" name is fine because the namespace is often qualifying it at the callsite. But for .NET, I would recommend calling this `ClientEnvConfig`. Also, same for `ClientEnvConfigProfile` and `ClientEnvConfigTls`, though I wonder if we should consider just making those two, `Profile` and `Tls`, as inner classes here. Not sure, thoughts?

cretz commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
```suggestion
            IReadOnlyDictionary<string, string>? overrideEnvVars = null)
```

Here and elsewhere

cretz commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
I don't think we need to publicly support JSON serializability of env config in any SDK

cretz commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
Make this a private static readonly field

cretz commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
Just a note, you can shorten these one-line methods:

```suggestion
        public string ToJson(JsonSerializerOptions? options = null) =>
            JsonSerializer.Serialize(this, options ?? GetDefaultJsonOptions());
```

cretz commented on src/Temporalio/Client/Configuration/ClientConfigProfile.cs:
Mentioned in other place, but this should be a `record` IMO

cretz commented on src/Temporalio/Client/Configuration/ClientConfigProfile.cs:
Pedantic, but this is more of an `ArgumentError`

cretz commented on src/Temporalio/Client/Configuration/ClientConfigProfile.cs:
General question, what happens if I provide a `configSource` with some data and set `disableFile` as `true`? Does it disregard my `configSource` parameter? Or if not, then it is reasonable to have `disableFile` and `disableEnv` when there is a `configSource` right?

cretz commented on src/Temporalio/Client/Configuration/ClientConfigProfile.cs:
```suggestion
        public TemporalClientConnectOptions ToClientConnectOptions()
```

cretz commented on src/Temporalio/Client/Configuration/ClientConfigTls.cs:
Mentioned elsewhere, but let's make this a `record`

cretz commented on src/Temporalio/Client/Configuration/DataSource.cs:
```suggestion
        public string? Path { get; private init; }

        /// <summary>
        /// Gets the raw data for this data source, if applicable.
        /// </summary>
        public byte[]? Data { get; private init; }
```

These should only be mutable on init

cretz commented on src/Temporalio/Client/Configuration/DataSource.cs:
```suggestion
        /// <param name="content">Configuration data as a UTF-8 string.</param>
```

cretz commented on src/Temporalio/Client/Configuration/DataSource.cs:
I don't understand this one exactly. What is the use case?

cretz commented on src/Temporalio/Client/Configuration/DataSource.cs:
Mentioned before, but I don't think JSON serializability makes sense for env config

cretz commented on src/Temporalio/Bridge/EnvConfig.cs:
Mentioned in Core review, the C bridge calls here don't need to be async IMO. That may change the structure enough where I'll hold off reviewing this file until that part changes.

THardy98 committed:
feat: Update bridge integration to use new synchronous C API

THardy98 committed:
Convert ClientConfig classes to records and improve immutability

- Convert ClientConfig, ClientConfigProfile, and ClientConfigTls to record types
- Remove ICloneable interface (records have built-in cloning)
- Use private init setters instead of private set
- Replace Dictionary parameters with IReadOnlyDictionary in public API

Replace JSON serialization with record serialization

- Replace ToJson()/FromJson() methods with ToRecord()/FromRecord() methods
- Create ClientConfigTlsRecord and ClientConfigProfileRecord for TOML structure
- Support conversion between config objects and TOML-ready record structures

Rename namespace from Configuration to EnvConfig

- Change namespace: Temporalio.Client.Configuration → Temporalio.Client.EnvConfig

Rename ClientConfig class to ClientEnvConfig

- Rename ClientConfig → ClientEnvConfig
- Update all references in tests and documentation
- Rename file ClientConfig.cs → ClientEnvConfig.cs
- Renaming EnvProfile -> ConfigProfile

Fix TLS disabled tri-state and profile not found behavior in environment config

- Change Tls.Disabled from bool to bool? to support tri-state behavior (null/true/false)
- Consolidate DataSource class back into ClientEnvConfig.cs
- Add test for tri-state TLS behavior and profile resolution
- Update profile loading logic to match Rust core behavior:
  - profile=null with missing "default" → return empty profile
  - profile="default" with missing "default" → throw exception

cretz commented on .claude/settings.local.json:
Can this file be removed?

cretz commented on CLAUDE.md:
Can this file be removed?

cretz reviewed (commented):
Is this ready for re-review?

THardy98 committed:
Remove CLAUDE files

THardy98 committed:
Remove useless disableFile parameter when loading ClientConfig. Parse Vec<u8> from DataSource instead of String

THardy98 commented on .claude/settings.local.json:
removed

THardy98 commented on CLAUDE.md:
removed

THardy98 commented:
> Is this ready for re-review?

Yes, RFR. Will resolve the above comments

THardy98 commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
Renamed

THardy98 commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
Yes - so this was intended to be the .NET equivalent of `to_dict/from_dict` in Python. Realized later this makes more sense as `toRecord/fromRecord` than JSON

THardy98 commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
removed

THardy98 commented on src/Temporalio/Client/Configuration/ClientConfig.cs:
Removed

THardy98 commented on src/Temporalio/Client/Configuration/ClientConfigProfile.cs:
Removed - we handle this case lower in core

THardy98 commented on src/Temporalio/Client/Configuration/ClientConfigProfile.cs:
If `disableFile` is `true`, then the `configSource` will be ignored. We return an empty profile object.

THardy98 commented on src/Temporalio/Client/Configuration/ClientConfigProfile.cs:
Here's the code snippet from core:
```
let mut profile = if options.disable_file {
        ClientConfigProfile::default()
    } else {
        // Load the full config
        let config = load_client_config(
            LoadClientConfigOptions {
                config_source: options.config_source,
                config_file_strict: options.config_file_strict,
            },
            env_vars,
        )?;
        ...
    }
```

THardy98 commented on src/Temporalio/Client/Configuration/DataSource.cs:
Removed - we want to propagate null data sources to core

THardy98 commented on src/Temporalio/Client/Configuration/DataSource.cs:
Removed (in favor of serializing to/from dictionary)

THardy98 commented on src/Temporalio/Bridge/EnvConfig.cs:
Yup - refactored

THardy98 committed:
Tests formatting

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Can probably just move all of this to the primary class constructor where you make the `record`. Same for other objects here.

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
I wouldn't necessarily call this a "record", maybe `ToDictionary`? Arguably across SDKs this would be a dictionary with a single "profiles" entry if we wanted to properly match TOML, but I see we didn't really do that anywhere.

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
```suggestion
        public static DataSource FromUTF8String(string content)
```

This would be a more .NET like name IMO

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Should mention it defaults to a default disk location IMO (don't have to get into detail on where/how). Was a bit confusing at first.

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
```suggestion
        public IReadOnlyDictionary<string, ProfileRecord> ToRecord() => Profiles.ToDictionary(
```

Pedantic, but remember there's a shortcut for single-line methods

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
We shouldn't be setting the `Domain` on TLS options, that is for SNI override. If we are doing that in other SDKs, we should fix it IMO. Address check does not need to be in the if, and this same logic should be removed from the TLS section.

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
We qualified "profile" as `ClientProfile` even though it is nested in `ClientEnvConfig`, I think we should either not qualify `ClientProfile` or qualify this as `ClientTls` (I think the former may make more sense).

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
I think this can be a method on the class instead of a static helper

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Pedantic, but IMO the caller can do the null check and this can always accept and always return non-null

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Hrmm, this doesn't seem quite right. So this is "profile name override" right and if left null, Core will use the `TEMPORAL_PROFILE` env var correct? But if you _always_ pass in this value to Core, how does it know to fall back to the env var? (may need to check other SDKs if we're doing this there too)

cretz commented on src/Temporalio/Bridge/EnvConfig.cs:
Not sure you have to be forgiving with your own internal JSON format

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Gonna sound a bit annoying, but lots of parameters like this is not that common in .NET. I think we should consider having a `ConfigProfileLoadOptions` (which can also be used in `LoadClientConnectOptions`), and to be consistent we should have a `ClientEnvConfigLoadOptions` for use with `ClientEnvConfig`. Can see other options classes to see how we've done those in .NET (basically cloneable, no required constructor params, all options optional).

cretz commented on src/Temporalio/Bridge/EnvConfig.cs:
Mentioned in Core PR, but I think we should use the same string-passing approach we have elsewhere, namely `scope.ByteArray(`. Same for other places in this file.

cretz commented on src/Temporalio/Bridge/EnvConfig.cs:
```suggestion
                        data = scope.ByteArray(source?.Data),
```

It's already built to handle nulls

cretz commented on src/Temporalio/Bridge/EnvConfig.cs:
Not sure switch expressions are supported in all versions of C# we support, haven't checked though

cretz commented on src/Temporalio/Bridge/EnvConfig.cs:
Which situations is this one vs the other? Doesn't make sense IMO to have a forgiving parser for our own internal JSON format (not _completely_ against the use of JSON to pass across Core, just a bit abnormal).

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Why a "record" instead of a dictionary? That record won't necessarily serialize to TOML correctly which is the purpose of this IIRC

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Oh this is wrong - I had removed this but I might have lost the change in a rebase. Fixing this

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
I had it as `Profile` originally but .NET complained that it had the same name as `System.something.Profile`, you couldn't name this as `Profile` as well. I opted for `ConfigProfile` because of that, open to `EnvProfile` or `ClientProfile` as well. I think they are all look the same amount of weird after `ClientEnvConfig`

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Unless there's a way to get around that, I'll rename `Tls` -> `ConfigTls` (or `Client<x>` or `Env<x>` whatever we prefer, I have no preference)

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Can't always return non-null, in the case where no `data` or `path` keys exist in the dictionary. It shouldn't happen in principle, but would rather guard this case than not. Given that, I don't really see much value in making it such that callers have to null check

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
> I had it as Profile originally but .NET complained that it had the same name as System.something.Profile, you couldn't name this as Profile as well. [...] Unless there's a way to get around that,

IMO you can silence this warning here since it's a nested class. Everyone will access this as `ClientEnvConfig.Profile` I think. Can you show the error here so I can check the MS page on it? Let me know if not obvious how to use pragmas to disable that check for this line (assuming it is disable-able).

THardy98 committed:
Address most PR feedback

THardy98 committed:
Use scope.ByteArray in favor of sbyte*

THardy98 committed:
Deserialize JSON from core using JSON DTO objects

THardy98 commented on src/Temporalio/Bridge/EnvConfig.cs:
Yeah - I was having trouble deserializing the incoming JSON properly (nested JSON) so had this as a fallback which worked. Not very nice though.

I've removed this in favor of actual JSON DTO objects, much cleaner.

THardy98 commented on src/Temporalio/Bridge/EnvConfig.cs:
Removed

THardy98 commented on src/Temporalio/Bridge/EnvConfig.cs:
removed

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Removed

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Not done in other SDKs, this is a code-gen artifact I missed. Removed all `Domain` setting logic. Thanks for the catch.

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Why is it the case that a record won't serialize to TOML correctly? I haven't seen anything on this.

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
This is noted in the summaries of `ConfigLoadOptions.DataSource` and `ProfileLoadOptions.DataSource`

THardy98 committed:
Clean up - renaming and using shorthand returns

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Depends on the TOML library in use on how it handles PascalCased properties and such, but even if it can, users will have to know to set the setting to convert to underscore-based names and such. If you give/accept a dictionary (of dictionaries/lists/primitives), you can control the key names. And as a bonus you don't have to have new classes.

THardy98 committed:
Use renamed core methods with Env prefix before Config. Use tri-state TLS configuration (disabled, enabled, not configured)

THardy98 committed:
Rename ConfigProfile -> Profile, with suppression of build warning

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Changed to `ToDictionary`.

wrt changing to a single "profiles" entry, if it's something we care about I can open some PRs once this is in

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Sweet - suppressed the warning. Much nicer.

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Have changed this to use dictionaries instead of records

THardy98 committed:
TlsRecord & ProfileRecord removed, in favor of to/fromDictionary methods on the existing Tls and Profile inner classes

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
```suggestion
#pragma warning disable CA1724 // We're ok with the Profile name since it's a nested class
        public sealed record Profile(
#pragma warning restore CA1724
```

I think this is the better/clearer way

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
```suggestion
        public static TemporalClientConnectOptions LoadClientConnectOptions(ProfileLoadOptions? options = null)
```

We should make the options parameter optional on all of these if the default `new()` is acceptable. Can do `options ??= new()` in the method just before use if needed to be non-null in code (or something like that). But I think we want our samples code to just be able to `TemporalClient.ConnectAsync(ClientEnvConfig.LoadClientConnectOptions())`

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Ok, sorry I am just noticing this. So we should not make this `.Default` call to runtime. It lazily creates a thread pool and does other things. The only reason runtime seems like it is needed is for freeing byte arrays. But that is just an added optional optimization I put in there if I wanted one day to come back and reuse byte array buffers. Not only is the `return_buf` operation not even implemented, but `temporal_core_byte_array_free` actually accepts a null pointer for runtime, which we should use in this case.

I only required runtimes for freeing byte arrays because I always had a runtime available. But this is the first use case where it doesn't make. I think we should:

1. Update `ByteArray` constructor/field to accept `Runtime?` instead of `Runtime`
2. Remove `Runtime.FreeByteArray` method altogether
3. Update `ByteArray.ReleaseHandle` to explicitly invoke `Interop.Methods.temporal_core_byte_array_free` with the first param as `Runtime?.Ptr` so it can be null
4. Update your instantiation of `ByteArray` to pass in `null` for the runtime

Sorry about just realizing this

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
```suggestion
            public Dictionary<string, object> ToDictionary() =>
```

Not sure any of these values in these dictionaries should be nullable. The key should be absent of the value is null IMO.

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
```suggestion
```

Default constructor is implied when there are no other constructors (same for other options class)

cretz reviewed (commented):
Few more things, but I think we're basically there

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
It is impossible to reach this line, so arguably this method should never return null (and therefore to my other comment, I wonder if we should never accept a null dictionary or return null data source from dictionary, but can leave that if you really want).

cretz commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Shouldn't have to use `this.` in these cases

THardy98 committed:
Make to/from dictionary conversion non-null. Minor style improvements

THardy98 committed:
Remove runtime usage

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
Removed `runtime` usage

THardy98 commented on src/Temporalio/Client/EnvConfig/ClientEnvConfig.cs:
options are nullable and use fallback defaults

THardy98 committed:
Update submodule with merged commit

THardy98 committed:
Merge branch 'main' into env_config

THardy98 committed:
Formatting

THardy98 committed:
Fix interop conflicts

THardy98 committed:
update test file paths for windows

THardy98 committed:
Merge branch 'main' into env_config
